src/web.ts

```diff
@@ -0,0 +1,532 @@
+  0% {
+    transform: rotate(0deg);
+  }
+  100% {
+    transform: rotate(360deg);
+  }
+}
+
+`
+
+@customElement("ad4m-connect")
+export default class Ad4mConnect extends LitElement {
+  static styles = [styles];
+
+  @state()
+  private _client = null;
+
+  @state()
+  private _state = null;
+
+  @state()
+  private _code = null;
+
+  @state()
+  private _url = null;
+
+  @property({ type: String, reflect: true })
+  appname = null
+
+  @property({ type: String, reflect: true })
+  appdesc = null
+
+  @property({ type: String, reflect: true })
+  appdomain = null
+
+  @property({ type: String, reflect: true })
+  capabilities
+
+  @property({ type: String, reflect: true })
+  token
+
+  @property({ type: String, reflect: true })
+  url
+
+  @property({ type: String, reflect: true })
+  port
+
+  @property({ type: String, reflect: true })
+  appiconpath
+
+  getClient() {
+    console.log('client', this._client)
+    return this._client;
+  }
+
+  connectedCallback() {
+    super.connectedCallback();
+
+    const link = document.createElement("link");
+    link.rel = "stylesheet";
+    link.type = "text/css";
+    link.crossOrigin = "anonymous";
+    link.href =
+      "https://fonts.googleapis.com/css2?family=Comfortaa:wght@300;400;700&display=swap";
+    document.head.appendChild(link);
+
+    const client = ad4mConnect({
+      appName: this.appname,
+      appDesc: this.appdesc,
+      appDomain: this.appdomain,
+      capabilities: JSON.parse(this.capabilities),
+      port: this.port,
+      token: this.token,
+      url: this.url
+    });
+
+    console.log(this.appname, this.appdesc, this.appdomain, JSON.parse(this.capabilities))
+
+    this._client = client;
+
+    client?.addEventListener('loading', () => {
+      console.log('loading');
+      this._state = 'loading';
+    })
+
+    client?.addEventListener('not_connected', () => {
+      console.log('not_connected');
+      this._state = 'not_connected';
+    })
+
+    client?.addEventListener('init', () => {
+      console.log('init');
+      this._state = 'init';
+    })
+
+    client?.addEventListener('capabilties_not_matched', (isFirst) => {
+      if (isFirst) {
+        this._state = 'capabilties_not_matched_first';
+      } else {
+        this._state = 'capabilties_not_matched';
+      }
+    })
+
+    client?.addEventListener('request_capability', (requestId) => {
+      console.log('request_capability', requestId);
+      this._state = 'request_capability';
+    })
+
+    client?.addEventListener('connected_with_capabilities', () => {
+      console.log('connected_with_capabilities');
+      this._state = 'connected_with_capabilities';
+    })
+
+    document.addEventListener('fetch-ad4m-client', () => {
+      console.log('lul')
+      const event = new CustomEvent('return-fetch-ad4m-client', { detail: this._client });
+      document.dispatchEvent(event);
+    });
+  }
+
+  render() {
+    const state = this._state;
+    if (state === 'loading' || state === 'na') {
```

Comment:
This seems a lot of complex, should we include some docs and screenshots for onboarding devs?

Reply:
And a bit concerned about how to maintain so many states, I'm thinking if it's possible to reduces the state, give them a long but consistent name format, and some unit tests.

Reply:
yes, I guess we should make states more explicit in naming, but I don't think we will be able to reduce the states more than this.

Reply:
also developers who just want to ingerate the the webcomponent or use electron don't need to be concerned with states, only people who want to build their own UI will need to be concerened.